agent/consul/grpc_resolver.go

```diff
@@ -0,0 +1,105 @@
+package consul
+
+import (
+	"sync"
+
+	"github.com/hashicorp/consul/agent/metadata"
+	"google.golang.org/grpc/resolver"
+)
+
+const grpcResolverScheme = "consul"
+
+var resolverBuilder *ServerResolverBuilder
+
+func init() {
+	resolverBuilder = NewServerResolverBuilder()
+	resolver.Register(resolverBuilder)
+}
+
+// ServerResolverBuilder tracks the current server list and keeps any
+// ServerResolvers updated when changes occur.
+type ServerResolverBuilder struct {
+	servers   map[string]*metadata.Server
+	resolvers map[*ServerResolver]struct{}
+	lock      sync.Mutex
+}
+
+func NewServerResolverBuilder() *ServerResolverBuilder {
+	return &ServerResolverBuilder{
+		servers:   make(map[string]*metadata.Server),
+		resolvers: make(map[*ServerResolver]struct{}),
+	}
+}
+
+func (s *ServerResolverBuilder) Build(target resolver.Target, cc resolver.ClientConn, opts resolver.BuildOption) (resolver.Resolver, error) {
```

Comment:
It's pretty hard for me to follow the code in this file in general. I've even poked around in gRPC resolvers before and thought I had a handle on how they worked but I'm struggling to understand what is going on here.

Could we maybe add doc comments to all the methods describing when they are called by which component and the significance of the logic in them?

e.g. is this builder just used one at setup or on every new RPC?